File: android/sdk/src/main/kotlin/com/clevertap/android/nativedisplay/handler/ActionHandler.kt
package com.clevertap.android.nativedisplay.handler

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import com.clevertap.android.nativedisplay.bridge.NativeDisplayBridge
import com.clevertap.android.nativedisplay.listener.NativeDisplayActionListener
import com.clevertap.android.nativedisplay.models.Action
import com.clevertap.android.nativedisplay.models.ExecutionMode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.long
import kotlinx.serialization.json.longOrNull
import com.clevertap.android.nativedisplay.listener.InteractionType
import com.clevertap.android.nativedisplay.listener.NativeDisplayComponentListener

/**
 * Handles execution of actions triggered by Native Display components.
 *
 * This class is responsible for:
 * - Routing actions to appropriate handlers
 * - Converting JSON data to usable Kotlin types
 * - Notifying the client listener
 * - Providing default implementations for certain actions (like opening URLs)
 * - Managing coroutine lifecycle for async operations
 *
 * @param context Android context for starting intents, opening URLs, etc.
 * @param listener Client's callback interface for handling actions
 * @param componentListener Optional component interaction listener
 * @param unitId The CleverTap display unit ID (`wzrk_id`) for attribution events; null when
 *        the config did not come from a Core SDK display unit payload
 * @param pushViewedEvent Test seam invoked when "Notification Viewed" fires. Production
 *        default forwards to [NativeDisplayBridge.pushViewedEvent], which is a no-op when
 *        no CleverTap Core SDK is wired — making the auto-attribution path safe regardless
 *        of whether a client listener is supplied.
 * @param pushClickedEvent Test seam invoked when "Notification Clicked" fires. Production
 *        default forwards to [NativeDisplayBridge.pushClickedEvent], same no-op guarantee.
 */
internal class ActionHandler(
    private val context: Context,
    private val listener: NativeDisplayActionListener?,
    private val componentListener: NativeDisplayComponentListener? = null,
    private val unitId: String? = null,
    private val pushViewedEvent: (String) -> Unit = DEFAULT_PUSH_VIEWED,
    private val pushClickedEvent: (String, Map<String, Any?>?) -> Unit = DEFAULT_PUSH_CLICKED,
) {

    private val coroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private val firedSystemEvents = mutableSetOf<String>()

    companion object {
        private const val TAG = "ActionHandler"

        /**
         * Default attribution forwarders — invoke the singleton bridge so that
         * `pushDisplayUnitViewedEventForID` / `pushDisplayUnitClickedEventForID` (or its
         * element-aware successor) fire on the CleverTap Core SDK whenever it is wired,
         * regardless of whether the host app supplied a [NativeDisplayActionListener].
         *
         * When Core SDK is absent the bridge's push methods short-circuit (return false) —
         * the auto-wire path is a graceful no-op.
         *
         * Viewed is unit-level (no extras); Clicked carries action extras when the host's
         * Core SDK exposes the element-aware method, else falls back to unit-level click.
         */
        private val DEFAULT_PUSH_VIEWED: (String) -> Unit = { unitId ->
            NativeDisplayBridge.getInstance()?.pushViewedEvent(unitId)
        }
        private val DEFAULT_PUSH_CLICKED: (String, Map<String, Any?>?) -> Unit = { unitId, extras ->
            NativeDisplayBridge.getInstance()?.pushClickedEvent(unitId, extras)
        }
    }

    /**
     * Execute an action based on its type.
     * This is the main entry point for action execution.
     *
     * @param action The action to execute
     * @param nodeId The ID of the node that triggered this action (for debugging)
     * @param interactionType The type of interaction that triggered this action
     */
    fun handleAction(
        action: Action,
        nodeId: String,
        interactionType: InteractionType = InteractionType.CLICK
    ) {
        coroutineScope.launch {
            try {
                Log.d(TAG, "Handling action for node: $nodeId, action: ${action::class.simpleName}")

                // Notify component listener first (if interested in this node)
                val shouldProceed = notifyComponentListener(
                    nodeId = nodeId,
                    interactionType = interactionType,
                    hasServerAction = true
                )

                // If component listener consumed the interaction, stop here
                if (!shouldProceed) {
                    Log.d(TAG, "Component listener consumed interaction for node: $nodeId")
                    return@launch
                }

                // Proceed with server action
                when (action) {
                    is Action.OpenUrl -> handleOpenUrl(action, nodeId)
                    is Action.CustomAction -> handleCustomAction(action, nodeId)
                    is Action.Navigate -> handleNavigate(action, nodeId)
                    is Action.TrackEvent -> handleTrackEvent(action, nodeId)
                    is Action.CompositeAction -> handleCompositeAction(action, nodeId)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error handling action for node: $nodeId", e)
                listener?.onActionError(action, e)
            }
        }
    }

    /**
     * Handle trigger for interactions where no server action is defined.
     * Still notifies component listener if they're interested.
     *
     * @param nodeId The ID of the component
     * @param interactionType The type of interaction
     */
    fun handleInteractionWithoutAction(
        nodeId: String,
        interactionType: InteractionType
    ) {
        coroutineScope.launch {
            try {
                Log.d(TAG, "Handling interaction without action for node: $nodeId")
                notifyComponentListener(
                    nodeId = nodeId,
                    interactionType = interactionType,
                    hasServerAction = false
                )
            } catch (e: Exception) {
                Log.e(TAG, "Error handling interaction for node: $nodeId", e)
            }
        }
    }

    /**
     * Notify component listener about an interaction.
     *
     * @return true if should proceed with server action, false if consumed by listener
     */
    private suspend fun notifyComponentListener(
        nodeId: String,
        interactionType: InteractionType,
        hasServerAction: Boolean
    ): Boolean {
        return withContext(Dispatchers.Main) {
            componentListener?.let { listener ->
                // Check if listener is interested in this node
                val interestedNodeIds = listener.getInterestedNodeIds()
                val isInterested = interestedNodeIds == null || interestedNodeIds.contains(nodeId)

                if (isInterested) {
                    // Call listener and check if it consumed the interaction
                    val consumed = listener.onComponentInteraction(
                        nodeId = nodeId,
                        interactionType = interactionType,
                        hasServerAction = hasServerAction
                    )

                    if (consumed) {
                        Log.d(TAG, "Component listener consumed interaction for: $nodeId")
                        return@withContext false  // Don't proceed with server action
                    }
                }
            }

            return@withContext true  // Proceed with server action
        }
    }

    /**
     * Execute multiple actions for a given trigger (e.g., "onClick").
     *
     * @param actions Map of trigger names to actions
     * @param trigger The trigger name (e.g., "onClick", "onLongPress")
     * @param nodeId The ID of the node that triggered this action
     */
    fun handleTrigger(
        actions: Map<String, Action>?,
        trigger: String,
        nodeId: String
    ) {
        val action = actions?.get(trigger) ?: return
        handleAction(action, nodeId)
    }

    /**
     * Execute a lifecycle action (onAppear/onDisappear).
     * These bypass the component listener since they are not user interactions.
     *
     * @param action The action to execute
     * @param nodeId The ID of the node that triggered this action
     */
    fun handleLifecycleAction(
        action: Action,
        nodeId: String
    ) {
        coroutineScope.launch {
            try {
                Log.d(TAG, "Handling lifecycle action for node: $nodeId, action: ${action::class.simpleName}")
                when (action) {
                    is Action.OpenUrl -> handleOpenUrl(action, nodeId)
                    is Action.CustomAction -> handleCustomAction(action, nodeId)
                    is Action.Navigate -> handleNavigate(action, nodeId)
                    is Action.TrackEvent -> handleTrackEvent(action, nodeId)
                    is Action.CompositeAction -> handleCompositeAction(action, nodeId)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error handling lifecycle action for node: $nodeId", e)
                listener?.onActionError(action, e)
            }
        }
    }

    /**
     * Fire a hardcoded system event through the action listener AND the CleverTap
     * Core SDK bridge (when wired).
     *
     * System events are SDK-level events that always fire (not server-driven).
     * Notification Viewed / Notification Clicked are also forwarded to the bridge
     * so attribution flows to Core SDK automatically whenever it is integrated —
     * the client does not have to opt in by attaching a listener.
     *
     * @param eventName The system event name (e.g., "Notification Viewed")
     * @param properties Optional event properties
     * @param deduplicate When true, skips firing if this event name was already fired
     *                    by this handler instance
     */
    fun fireSystemEvent(eventName: String, properties: Map<String, Any?>? = null, deduplicate: Boolean = false) {
        if (deduplicate && !firedSystemEvents.add(eventName)) {
            Log.d(TAG, "System event already fired, skipping: $eventName")
            return
        }
        coroutineScope.launch {
            try {
                Log.d(TAG, "Firing system event: $eventName")
                listener?.onTrackEvent(eventName, properties)
                if (unitId != null) {
                    when (eventName) {
                        "Notification Viewed" -> {
                            listener?.onDisplayUnitViewed(unitId)
                            pushViewedEvent(unitId)
                        }
                        "Notification Clicked" -> {
                            listener?.onDisplayUnitClicked(unitId)
                            pushClickedEvent(unitId, properties)
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error firing system event: $eventName", e)
            }
        }
    }

    /**
     * Handle OPEN_URL action.
     * First checks if client wants to handle it, then uses default behavior.
     */
    private suspend fun handleOpenUrl(action: Action.OpenUrl, nodeId: String) {
        withContext(Dispatchers.Main) {
            Log.d(TAG, "Opening URL: ${action.url} (openInBrowser: ${action.openInBrowser})")

            // Ask listener if they want to handle it
            val handled = listener?.onOpenUrl(
                action.url,
                action.openInBrowser
            ) ?: false

            // If listener didn't handle it, use default behavior
            if (!handled) {
                executeDefaultOpenUrl(action)
            }
        }
    }

    private fun executeDefaultOpenUrl(action: Action.OpenUrl) {
        try {
            if (action.customTabsEnabled) {
                openInCustomTab(action.url)
                return
            }
            openUrl(action.url)
        } catch (_: Exception) {
            Log.d(TAG, "No activity found to open url: ${action.url}")
            listener?.onActionError(action, Exception("No activity found to open url: ${action.url}"))
        }
    }

    // Matches Core SDK InAppActionHandler.openUrl behavior exactly.
    private fun openUrl(url: String) {
        val uri = Uri.parse(url.replace("\n", "").replace("\r", ""))
        val queryBundle = Bundle()
        uri.queryParameterNames?.forEach { name ->
            queryBundle.putString(name, uri.getQueryParameter(name))
        }
        val intent = Intent(Intent.ACTION_VIEW, uri).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
            if (!queryBundle.isEmpty) putExtras(queryBundle)
        }
        val appPackageName = context.packageName
        context.packageManager.queryIntentActivities(intent, 0).forEach { resolveInfo ->
            if (appPackageName == resolveInfo.activityInfo.packageName) {
                intent.setPackage(appPackageName)
                return@forEach
            }
        }
        context.startActivity(intent)
        Log.d(TAG, "Opened URL: $url")
    }

    private fun openInCustomTab(url: String) {
        openUrl(url)
    }

    /**
     * Handle CUSTOM action.
     * Converts JSON value to Kotlin types and notifies listener.
     */
    private suspend fun handleCustomAction(action: Action.CustomAction, nodeId: String) {
        withContext(Dispatchers.Main) {
            Log.d(TAG, "Executing custom action: ${action.key}")

            val parsedValue = parseJsonValue(action.value)

            listener?.onCustomAction(
                key = action.key,
                value = parsedValue,
                metadata = action.metadata
            )
        }
    }

    /**
     * Handle NAVIGATE action.
     * Notifies listener to perform navigation.
     */
    private suspend fun handleNavigate(action: Action.Navigate, nodeId: String) {
        withContext(Dispatchers.Main) {
            Log.d(TAG, "Navigating to: ${action.destination}")

            listener?.onNavigate(
                destination = action.destination,
                params = action.params
            )
        }
    }

    /**
     * Handle TRACK_EVENT action.
     * Converts JSON properties to Kotlin types and notifies listener.
     */
    private suspend fun handleTrackEvent(action: Action.TrackEvent, nodeId: String) {
        withContext(Dispatchers.Main) {
            Log.d(TAG, "Tracking event: ${action.eventName}")

            val parsedProperties = action.properties?.mapValues { (_, value) ->
                parseJsonValue(value)
            }

            listener?.onTrackEvent(
                eventName = action.eventName,
                properties = parsedProperties
            )
        }
    }

    /**
     * Handle COMPOSITE action.
     * Executes multiple actions either sequentially or in parallel.
     */
    private suspend fun handleCompositeAction(action: Action.CompositeAction, nodeId: String) {
        Log.d(TAG, "Executing composite action with ${action.actions.size} sub-actions (${action.executionMode})")

        when (action.executionMode) {
            ExecutionMode.SEQUENTIAL -> {
                // Execute one after another
                action.actions.forEach { subAction ->
                    handleAction(subAction, "$nodeId-composite")
                }
            }
            ExecutionMode.PARALLEL -> {
                // Execute all at once
                action.actions.map { subAction ->
                    coroutineScope.async {
                        handleAction(subAction, "$nodeId-composite")
                    }
                }.awaitAll()
            }
        }
    }

    /**
     * Convert JsonElement to usable Kotlin types.
     *
     * Conversions:
     * - JsonPrimitive (string) → String
     * - JsonPrimitive (number) → Int/Long/Double
     * - JsonPrimitive (boolean) → Boolean
     * - JsonObject → Map<String, Any?>
     * - JsonArray → List<Any?>
     *
     * @param element The JSON element to parse
     * @return Parsed Kotlin value (String, Number, Boolean, Map, List, or null)
     */
    private fun parseJsonValue(element: JsonElement): Any? {
        return when (element) {
            is JsonPrimitive -> {
                when {
                    element.isString -> element.content
                    element.booleanOrNull != null -> element.boolean
                    element.intOrNull != null -> element.int
                    element.longOrNull != null -> element.long
                    element.doubleOrNull != null -> element.double
                    else -> element.content
                }
            }
            is JsonObject -> element.toMap()
            is JsonArray -> element.toList()
            else -> null
        }
    }

    /**
     * Convert JsonObject to Map<String, Any?>
     */
    private fun JsonObject.toMap(): Map<String, Any?> {
        return this.mapValues { (_, value) ->
            parseJsonValue(value)
        }
    }

    /**
     * Convert JsonArray to List<Any?>
     */
    private fun JsonArray.toList(): List<Any?> {
        return this.map { element ->
            parseJsonValue(element)
        }
    }

    /**
     * Clean up resources when ActionHandler is no longer needed.
     * Call this when the composable is disposed.
     */
    fun cleanup() {
        Log.d(TAG, "Cleaning up ActionHandler")
        coroutineScope.cancel()
    }
}

File: ios/Sources/CleverTapNativeDisplay/Handlers/ActionHandler.swift
//
//  ActionHandler.swift
//  CleverTapNativeDisplay
//
//  Created by Lalitkumar Patil on 05/01/26.
//


// MARK: - Action Handler
// Core action execution engine

import Foundation
import SafariServices
import UIKit

/// Handles execution of actions triggered by Native Display components.
class ActionHandler {

    private weak var actionListener: NativeDisplayActionListener?
    private weak var componentListener: NativeDisplayComponentListener?
    private var firedSystemEvents = Set<String>()
    private let unitId: String?

    /// Whether an action listener is attached (for callers to skip unnecessary work)
    var hasActionListener: Bool { actionListener != nil }

    init(
        actionListener: NativeDisplayActionListener?,
        componentListener: NativeDisplayComponentListener?,
        unitId: String? = nil
    ) {
        self.actionListener = actionListener
        self.componentListener = componentListener
        self.unitId = unitId
    }
    
    // MARK: - Public Methods
    
    /// Execute an action based on its type.
    /// - Parameters:
    ///   - action: The action to execute
    ///   - nodeId: The ID of the node that triggered this action
    ///   - interactionType: The type of interaction that triggered this action
    func handleAction(
        _ action: Action,
        nodeId: String,
        interactionType: InteractionType = .click
    ) {
        Task { @MainActor in
            do {
                print("ActionHandler: Handling action for node: \(nodeId)")
                
                // Notify component listener first (if interested in this node)
                let shouldProceed = notifyComponentListener(
                    nodeId: nodeId,
                    interactionType: interactionType,
                    hasServerAction: true
                )
                
                // If component listener consumed the interaction, stop here
                guard shouldProceed else {
                    print("ActionHandler: Component listener consumed interaction for node: \(nodeId)")
                    return
                }
                
                // Proceed with server action
                switch action {
                case .openUrl(let openUrlAction):
                    try await handleOpenUrl(openUrlAction, nodeId: nodeId)
                case .custom(let customAction):
                    handleCustomAction(customAction, nodeId: nodeId)
                case .navigate(let navigateAction):
                    handleNavigate(navigateAction, nodeId: nodeId)
                case .trackEvent(let trackEventAction):
                    handleTrackEvent(trackEventAction, nodeId: nodeId)
                case .composite(let compositeAction):
                    try await handleCompositeAction(compositeAction, nodeId: nodeId)
                }
            } catch {
                print("ActionHandler: Error handling action for node: \(nodeId), error: \(error)")
                actionListener?.onActionError(action: action, error: error)
            }
        }
    }
    
    /// Execute a lifecycle action (onAppear/onDisappear).
    /// These bypass the component listener since they are not user interactions.
    /// - Parameters:
    ///   - action: The action to execute
    ///   - nodeId: The ID of the node that triggered this action
    func handleLifecycleAction(
        _ action: Action,
        nodeId: String
    ) {
        Task { @MainActor in
            do {
                print("ActionHandler: Handling lifecycle action for node: \(nodeId)")
                switch action {
                case .openUrl(let openUrlAction):
                    try await handleOpenUrl(openUrlAction, nodeId: nodeId)
                case .custom(let customAction):
                    handleCustomAction(customAction, nodeId: nodeId)
                case .navigate(let navigateAction):
                    handleNavigate(navigateAction, nodeId: nodeId)
                case .trackEvent(let trackEventAction):
                    handleTrackEvent(trackEventAction, nodeId: nodeId)
                case .composite(let compositeAction):
                    try await handleCompositeAction(compositeAction, nodeId: nodeId)
                }
            } catch {
                print("ActionHandler: Error handling lifecycle action for node: \(nodeId) - \(error)")
                actionListener?.onActionError(action: action, error: error)
            }
        }
    }

    /// Fire a hardcoded system event through the action listener.
    /// System events are SDK-level events that always fire (not server-driven).
    /// - Parameters:
    ///   - eventName: The system event name (e.g., "Notification Viewed")
    ///   - properties: Optional event properties
    func fireSystemEvent(eventName: String, properties: [String: Any]? = nil, deduplicate: Bool = false) {
        guard hasActionListener else { return }
        if deduplicate {
            guard firedSystemEvents.insert(eventName).inserted else {
                print("ActionHandler: System event already fired, skipping: \(eventName)")
                return
            }
        }
        Task { @MainActor in
            print("ActionHandler: Firing system event: \(eventName)")
            actionListener?.onTrackEvent(eventName: eventName, properties: properties)
            if let unitId = unitId {
                switch eventName {
                case "Notification Viewed":
                    actionListener?.onDisplayUnitViewed?(unitId: unitId)
                case "Notification Clicked":
                    actionListener?.onDisplayUnitClicked?(unitId: unitId)
                default:
                    break
                }
            }
        }
    }

    /// Handle interaction for components without server actions.
    /// - Parameters:
    ///   - nodeId: The ID of the component
    ///   - interactionType: The type of interaction
    func handleInteractionWithoutAction(
        nodeId: String,
        interactionType: InteractionType
    ) {
        Task { @MainActor in
            print("ActionHandler: Handling interaction without action for node: \(nodeId)")
            _ = notifyComponentListener(
                nodeId: nodeId,
                interactionType: interactionType,
                hasServerAction: false
            )
        }
    }
    
    // MARK: - Private Helper Methods
    
    /// Notify component listener about an interaction.
    /// - Returns: true if should proceed with server action, false if consumed by listener
    @MainActor
    private func notifyComponentListener(
        nodeId: String,
        interactionType: InteractionType,
        hasServerAction: Bool
    ) -> Bool {
        guard let listener = componentListener else {
            return true // No listener, proceed with server action
        }
        
        // Check if listener is interested in this node
        let interestedNodeIds = listener.getInterestedNodeIds()
        let isInterested = interestedNodeIds == nil || interestedNodeIds?.contains(nodeId) == true
        
        guard isInterested else {
            return true // Not interested, proceed with server action
        }
        
        // Call listener and check if it consumed the interaction
        let consumed = listener.onComponentInteraction(
            nodeId: nodeId,
            interactionType: interactionType,
            hasServerAction: hasServerAction
        )
        
        if consumed {
            print("ActionHandler: Component listener consumed interaction for: \(nodeId)")
        }
        
        return !consumed // Return true if NOT consumed (should proceed)
    }
    
    // MARK: - Action Handlers
    
    @MainActor
    private func handleOpenUrl(_ action: Action.OpenUrlAction, nodeId: String) async throws {
        print("ActionHandler: Opening URL: \(action.url)")
        
        // Ask listener if they want to handle it
        let handled = actionListener?.onOpenUrl(
            url: action.url,
            openInBrowser: action.openInBrowser
        ) ?? false
        
        // If listener didn't handle it, use default behavior
        guard !handled else {
            return
        }
        
        // Execute default URL opening
        try executeDefaultOpenUrl(action)
    }
    
    @MainActor
    private func executeDefaultOpenUrl(_ action: Action.OpenUrlAction) throws {
        guard let url = URL(string: action.url) else {
            throw ActionError.invalidUrl(action.url)
        }
        
        // Validate URL scheme
        guard isValidUrlScheme(url.scheme) else {
            throw ActionError.invalidUrlScheme(url.scheme ?? "none")
        }
        
        if action.openInBrowser {
            // Open in external browser
            UIApplication.shared.open(url)
        } else if action.customTabsEnabled {
            // Open in SFSafariViewController (iOS equivalent of Chrome Custom Tabs)
            openInSafariViewController(url)
        } else {
            // Fallback to external browser
            UIApplication.shared.open(url)
        }
    }
    
    @MainActor
    private func openInSafariViewController(_ url: URL) {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
            // Fallback to external browser if we can't get root view controller
            UIApplication.shared.open(url)
            return
        }
        
        let safariVC = SFSafariViewController(url: url)
        safariVC.modalPresentationStyle = .pageSheet
        rootViewController.present(safariVC, animated: true)
    }
    
    private func isValidUrlScheme(_ scheme: String?) -> Bool {
        guard let scheme = scheme?.lowercased() else { return false }
        return ["http", "https", "tel", "mailto"].contains(scheme)
    }
    
    @MainActor
    private func handleCustomAction(_ action: Action.CustomAction, nodeId: String) {
        print("ActionHandler: Executing custom action: \(action.key)")
        
        let parsedValue = parseAnyCodableValue(action.value)
        
        actionListener?.onCustomAction(
            key: action.key,
            value: parsedValue,
            metadata: action.metadata
        )
    }
    
    @MainActor
    private func handleNavigate(_ action: Action.NavigateAction, nodeId: String) {
        print("ActionHandler: Navigating to: \(action.destination)")
        
        actionListener?.onNavigate(
            destination: action.destination,
            params: action.params
        )
    }
    
    @MainActor
    private func handleTrackEvent(_ action: Action.TrackEventAction, nodeId: String) {
        print("ActionHandler: Tracking event: \(action.eventName)")
        
        let parsedProperties = action.properties?.mapValues { parseAnyCodableValue($0) }
        
        actionListener?.onTrackEvent(
            eventName: action.eventName,
            properties: parsedProperties
        )
    }
    
    @MainActor
    private func handleCompositeAction(_ action: Action.CompositeAction, nodeId: String) async throws {
        print("ActionHandler: Executing composite action with \(action.actions.count) sub-actions (\(action.executionMode))")
        
        switch action.executionMode {
        case .sequential:
            // Execute one after another
            for subAction in action.actions {
                try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
                    handleAction(subAction, nodeId: "\(nodeId)-composite", interactionType: .click)
                    continuation.resume()
                }
            }
            
        case .parallel:
            // Execute all at once
            await withTaskGroup(of: Void.self) { group in
                for subAction in action.actions {
                    group.addTask {
                        self.handleAction(subAction, nodeId: "\(nodeId)-composite", interactionType: .click)
                    }
                }
            }
        }
    }
    
    // MARK: - Value Parsing
    
    /// Convert AnyCodable to usable Swift types.
    private func parseAnyCodableValue(_ anyCodable: AnyCodable) -> Any? {
        let value = anyCodable.value
        
        if value is NSNull {
            return nil
        } else if let bool = value as? Bool {
            return bool
        } else if let int = value as? Int {
            return int
        } else if let double = value as? Double {
            return double
        } else if let string = value as? String {
            return string
        } else if let array = value as? [Any] {
            return array.map { item in
                if let anyCodableItem = item as? AnyCodable {
                    return parseAnyCodableValue(anyCodableItem)
                }
                return item
            }
        } else if let dict = value as? [String: Any] {
            return dict.mapValues { item in
                if let anyCodableItem = item as? AnyCodable {
                    return parseAnyCodableValue(anyCodableItem)
                }
                return item
            }
        }
        
        return value
    }
}

// MARK: - Action Errors

enum ActionError: LocalizedError {
    case invalidUrl(String)
    case invalidUrlScheme(String)
    
    var errorDescription: String? {
        switch self {
        case .invalidUrl(let url):
            return "Invalid URL: \(url)"
        case .invalidUrlScheme(let scheme):
            return "Invalid URL scheme: \(scheme)"
        }
    }
}
